File: submissions/0020-valid-parentheses/solution.java
class Solution {
    
    List<Character> openBrackets = Arrays.asList(new Character[]{'(', '{', '['});
    List<Character> closeBrackets = Arrays.asList(new Character[]{')', '}', ']'});
    
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            Character c = s.charAt(i);
            if (openBrackets.contains(c)) {
                stack.push(c);
            } else { //close brackets
                
                if (stack.isEmpty()) return false;
                
                int index = closeBrackets.indexOf(c);
                Character expectedChar = openBrackets.get(index);
                Character actualChar = stack.pop();
                
                if (!actualChar.equals(expectedChar)) return false;
            }
            
        }
        
        return stack.size() == 0;
    }
}

File: submissions/0020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
    
        bracket_map = {
            '}' : '{',
            ')' : '(',
            ']' : '['
        }
        
        stack = []

        for c in s:
            if c in bracket_map.values():
                stack.append(c)
            else:
                if (len(stack) == 0):
                    return False
                check = stack.pop()
                if (check != bracket_map[c]):
                    return False
                
        
        return len(stack) == 0
        
